rpc/dial.go

```diff
@@ -99,85 +105,121 @@ func dial(
 		dOpts.webrtcOpts.Disable = true
 		dOpts.insecure = true
 		dOpts.disableDirect = false
+	} else if strings.ContainsRune(address, ':') {
+		isJustDomain = false
 	} else {
-		if strings.ContainsRune(address, ':') {
-			isJustDomain = false
-		} else {
-			isJustDomain = net.ParseIP(address) == nil
-		}
+		isJustDomain = net.ParseIP(address) == nil
 	}
 
+	connCh := make(chan dialSuccess, 2)
+	errCh := make(chan error, 2)
+	var dialAttempts int
 	if !dOpts.mdnsOptions.Disable && tryLocal && isJustDomain {
-		conn, cached, err := dialMulticastDNS(ctx, address, logger, dOpts)
-		if err != nil {
-			logger.Warnw("error dialing with mDNS; falling back to other methods", "error", err)
-		}
+		dialAttempts++
+		go func() {
+			conn, cached, err := dialMulticastDNS(ctx, address, logger, dOpts)
+			if err != nil {
+				logger.Warnw("error dialing with mDNS; falling back to other methods", "error", err)
+				errCh <- nil
+				return
+			}
 
-		if conn != nil {
-			return conn, cached, err
-		}
+			if conn != nil {
+				connCh <- dialSuccess{conn, cached}
+				return
+			}
+
+			errCh <- nil
+		}()
 	}
 
 	if !dOpts.webrtcOpts.Disable {
-		signalingAddress := dOpts.webrtcOpts.SignalingServerAddress
-		if signalingAddress == "" || dOpts.webrtcOpts.AllowAutoDetectAuthOptions {
-			if signalingAddress == "" {
-				// try WebRTC at same address
-				signalingAddress = address
-			}
-			target, port, err := getWebRTCTargetFromAddressWithDefaults(signalingAddress)
-			if err != nil {
-				return nil, false, err
```

Comment:
ditto https://github.com/viamrobotics/goutils/pull/234/files#r1449060104

Reply:
ya not sure about this - do the errors from getWebRTCTargetFromAddressWithDefaults seem like they would break direct grpc as well?

Reply:
I made [a follow-up ticket](https://viam.atlassian.net/browse/RSDK-6493) to investigate or possibly refactor this